File: 2020/day03/day03.hs
#!/usr/bin/env stack
-- stack --resolver lts-14.16 script

import System.Environment (getArgs)
import qualified Data.Vector as V

type TreeGrid = V.Vector (V.Vector Bool)

parseTrees :: String -> TreeGrid
parseTrees = V.fromList . map (V.fromList . map (=='#')) . lines

countTrees :: TreeGrid -> (Int, Int) -> Int
countTrees grid (dx, dy) = go (0, 0) 0
  where go (x, y) acc
          | y >= V.length grid = acc
          | otherwise          =
              let row    = grid V.! y
                  isTree = row V.! (x `mod` V.length row)
                  acc' = if isTree then acc + 1 else acc
              in go (x + dx, y + dy) acc'

star1 :: TreeGrid -> Int
star1 = flip countTrees (3, 1)

star2 :: TreeGrid -> Int
star2 grid = product $ map (countTrees grid) $ angles
  where angles = [(1, 1), (3, 1), (5, 1), (7, 1), (1, 2)]

main :: IO ()
main = do
  args <- getArgs
  let
    filename = case args of
      (x:_) -> x
      _     -> "input.txt"
  input <- parseTrees <$> readFile filename
  print $ star1 input
  print $ star2 input

File: 2020/day03/day03.rs
#!/usr/bin/env run-cargo-script

fn count_trees(grid: &Vec<Vec<bool>>, angle: (usize, usize)) -> usize {
    let mut count = 0;
    let (mut x, mut y) = (0, 0);
    let (dx, dy) = angle;
    while y < grid.len() {
        let row = &grid[y];
        if row[x % row.len()] {
            count += 1;
        }
        x += dx;
        y += dy;
    }

    count
}

fn star1(input: &Vec<Vec<bool>>) -> usize {
    count_trees(input, (3, 1))
}

fn star2(input: &Vec<Vec<bool>>) -> usize {
    let angles = &[(1, 1), (3, 1), (5, 1), (7, 1), (1, 2)];
    angles
        .iter()
        .map(|i| count_trees(input, *i))
        .fold(1, |x, y| x * y)
}

fn main() {
    let args: Vec<_> = std::env::args().collect();
    let filename = args.get(1).map(|s| &s[..]).unwrap_or("input.txt");
    let input = std::fs::read_to_string(filename)
        .unwrap()
        .lines()
        .map(|l| l.chars().map(|c| c == '#').collect())
        .collect();

    println!("{}", star1(&input));
    println!("{}", star2(&input));
}
